File: unification.cpp
#include "zion.h"
#include "dbg.h"
#include "logger.h"
#include <sstream>
#include "utils.h"
#include "types.h"
#include "unification.h"

unification_t::unification_t(
		bool result,
		std::string reasons,
		types::type_t::map bindings) :
	result(result),
	reasons(reasons),
	bindings(bindings)
{
	debug_above(10, log(log_info, "unification result {%s, %s, %s}",
				result ? "success" : "failure", reasons.c_str(),
				::str(bindings).c_str()));
}

types::type_t::ref prune(types::type_t::ref t, const types::type_t::map &bindings) {
	/* Follow the links across the bindings to reach the final binding. */
	std::string type_variable_name;
	if (get_type_variable_name(t, type_variable_name)) {
		auto binding_iter = bindings.find(type_variable_name);
        if (binding_iter != bindings.end()) {
            return prune(binding_iter->second, bindings);
		}
	}

	return t;
}

template <typename T>
bool occurs_in(const ptr<types::type_variable_t> &var, const T &types) {
    /* checks whether a type variable occurs in any other types. */
	for (auto &type : types) {
		if (occurs_in_type(var, type)) {
			return true;
		}
	}
	return false;
}

bool occurs_in_type(
		types::type_variable_t::ref var,
	   	types::type_t::ref b,
	   	types::type_t::map bindings)
{
	/* checks whether a type variable occurs in a type expression. must be
	 * called with var already pruned */
	assert(var == prune(var, bindings));
    auto pruned_b = prune(b, bindings);

	if (auto var2 = dyncast<const types::type_variable_t>(pruned_b)) {
		return var2 == var;
	} else if (auto type_operator = dyncast<const types::type_operator_t>(pruned_b)) {
		return occurs_in_type(var, type_operator->oper, bindings) ||
			occurs_in_type(var, type_operator->operand, bindings);
	} else {
		// TODO: handle type_product, type_sum
		return false;
	}
}

unification_t unify(
		types::type_t::ref lhs,
		types::type_t::ref rhs,
		types::type_t::map env,
		types::type_t::map bindings,
		int depth)
{
	if (depth > 120) {
		log(log_error, "unification depth is getting big...");
		dbg();
	}

	assert(lhs != nullptr);
	assert(rhs != nullptr);

	INDENT(7,
			string_format("unify(%s, %s, ..., %s)",
				lhs->str().c_str(),
				rhs->str().c_str(),
				str(bindings).c_str()));

	auto a = prune(lhs, bindings);
	auto b = prune(rhs, bindings);

	auto ptm_a = dyncast<const types::type_maybe_t>(a);
	auto ptm_b = dyncast<const types::type_maybe_t>(b);

	auto pti_a = dyncast<const types::type_id_t>(a);
	auto pti_b = dyncast<const types::type_id_t>(b);

	auto pto_a = dyncast<const types::type_operator_t>(a);
	auto pto_b = dyncast<const types::type_operator_t>(b);

	auto pts_a = dyncast<const types::type_sum_t>(a);
	auto pts_b = dyncast<const types::type_sum_t>(b);

	auto ptr_a = dyncast<const types::type_ptr_t>(a);
	auto ptr_b = dyncast<const types::type_ptr_t>(b);

	auto ptI_a = dyncast<const types::type_integer_t>(a);
	auto ptI_b = dyncast<const types::type_integer_t>(b);

	auto ptp_a = dyncast<const types::type_product_t>(a);

	auto ptf_a = dyncast<const types::type_function_t>(a);

	if (pti_a != nullptr) {
		/* check for basic type_id matching */
		if (pti_b != nullptr &&
				pti_a->id->get_name() == pti_b->id->get_name())
		{
			/* simple type_id match */
			return {true, "", bindings};
		} else if (pti_a->id->get_name() == "void") {
			/* everything is covariant with void because void encompasses anything (since it cannot be accessed and does
			 * not have a known size) */
			if (auto ptv_b = dyncast<const types::type_variable_t>(b)) {
				// TODO: probably we need to map all ftvs in b to void, not just the outer b
				bindings[ptv_b->id->get_name()] = a;
			}
			return {true, "", bindings};
		}
	}

	if (ptI_a != nullptr) {
	   	if (ptI_b == nullptr) {
			ptI_b = dyncast<const types::type_integer_t>(full_eval(b, env));
		}

		if (ptI_b != nullptr) {
			auto bit_size_unification = unify(ptI_a->bit_size, ptI_b->bit_size, env, bindings, depth + 1);
			if (bit_size_unification.result) {
				return {true, "", bit_size_unification.bindings};
			} else {
				return {
					false,
						string_format("bit-sizes did not match on %s and %s", a->str().c_str(), b->str().c_str()),
						bindings};
			}
		}
	}

	if (auto ptl_a = dyncast<const types::type_literal_t>(a)) {
		if (auto ptl_b = dyncast<const types::type_literal_t>(b)) {
			return {
				(ptl_a->token.text == ptl_b->token.text
				 && ptl_a->token.tk == ptl_b->token.tk),
				"", bindings};
		}
	}

	if (auto ptv_a = dyncast<const types::type_variable_t>(a)) {
		if (a != b) {
			if (occurs_in_type(ptv_a, b, bindings)) {
				return {
					false,
					string_format("recursive unification on %s and %s",
							a->str().c_str(), b->str().c_str()),
					bindings};
			}
			debug_above(4, log(log_info,
						"binding type_variable " c_id("%s") " to " c_type("%s"),
						ptv_a->id->get_name().c_str(),
						b->str(bindings).c_str()));
			assert(bindings.find(ptv_a->id->get_name()) == bindings.end());
			if (b->rebind(bindings)->ftv_count() != 0) {
				debug_above(4, log(log_info,
							"note that %s is itself not fully bound", b->str().c_str()));
			}
			bindings[ptv_a->id->get_name()] = b;
		}

		return {true, "", bindings};
	} else if (auto ptv_b = dyncast<const types::type_variable_t>(b)) {
		debug_above(7, log("flipping type variable from rhs to lhs"));
		return unify(ptv_b, a, env, bindings, depth + 1);
	} else if (ptm_a != nullptr) {
		if (ptm_b != nullptr) {
			debug_above(7, log("matching maybe types"));
			return unify(ptm_a->just, ptm_b->just, env, bindings, depth + 1);
		} else if (b->is_nil()) {
			debug_above(7, log("matching nil"));
			return {true, "", bindings};
		} else {
			debug_above(7, log("matching maybe on the lhs"));
			return unify(ptm_a->just, b, env, bindings, depth + 1);
		}
	} else if (pti_a != nullptr) {
		/* ok, we've got a mismatch, but we know we have an id on the left-hand
		 * side, let's try expanding the type_id to see whether it will unify
		 * after evaluation. */
		auto new_a = eval_id(pti_a, env);
		if (new_a != nullptr) {
			debug_above(6, log(log_info, "eval_id(%s, env) -> %s",
						pti_a->str().c_str(),
						new_a->str().c_str()));
			return unify(new_a, b, env, bindings, depth + 1);
		} else {
			/* types don't match */
			return {
				false,
				string_format("(%s != %s) and (%s could not be expanded further)",
						a->str().c_str(),
						b->str().c_str(),
						a->str().c_str()),
				{}};
		}
	} else if (ptp_a != nullptr) {
		if (auto ptp_b = dyncast<const types::type_product_t>(b)) {
			if (ptp_a->get_pk() != ptp_b->get_pk()) {
				return {
					false,
					string_format("product kinds are different %s <> %s (%s != %s)",
							pkstr(ptp_a->get_pk()),
							pkstr(ptp_b->get_pk()),
							ptp_a->str().c_str(),
							ptp_b->str().c_str()),
					bindings};
			}
			auto a_dimensions = ptp_a->get_dimensions();
			auto b_dimensions = ptp_b->get_dimensions();
			if (a_dimensions.size() != b_dimensions.size()) {
				return {
					false,
					string_format("product type lengths do not match "
							"(a = %s, b = %s)", ptp_a->str().c_str(),
							ptp_b->str().c_str()),
					bindings};
			} else {
				auto a_dims_end = a_dimensions.end();
				auto b_dims_iter = b_dimensions.begin();
				for (auto a_dims_iter = a_dimensions.begin();
						a_dims_iter != a_dims_end;
						++a_dims_iter, ++b_dims_iter) {
					debug_above(7, log("matching subitem in product type"));
					auto unification = unify(*a_dims_iter, *b_dims_iter,
							env, bindings, depth + 1);
					if (!unification.result) {
						return {false, unification.reasons, {}};
					}
					bindings = unification.bindings;
				}

				return {true, "products match", bindings};
			}
		} else {
			return {
				false,
				string_format("%s <> %s",
						a->str().c_str(),
						b->str().c_str()),
				bindings};
		}
	} else if (ptf_a != nullptr) {
		if (auto ptf_b = dyncast<const types::type_function_t>(b)) {
			debug_above(7, log("matching function arguments"));
			/* now make sure the arguments unify */
			auto args_unification = unify(ptf_a->args, ptf_b->args,
					env, bindings, depth + 1);
			if (!args_unification.result) {
				return {false, args_unification.reasons, {}};
			}
			bindings = args_unification.bindings;

			debug_above(7, log("matching function return types"));
			/* finally, make sure the return types unify */
			auto return_type_unification = unify(ptf_a->return_type, ptf_b->return_type,
					env, bindings, depth + 1);
			if (!return_type_unification.result) {
				return {false, return_type_unification.reasons, {}};
			}
			bindings = return_type_unification.bindings;
			return {true, "functions match", bindings};
		} else {
			return {
				false,
				string_format("%s <> %s",
						a->str().c_str(),
						b->str().c_str()),
				bindings};
		}
	} else if (pts_a != nullptr) {
		if (pts_b == nullptr) {
			std::vector<std::string> reasons;
			for (auto option : pts_a->options) {
				debug_above(7, log("matching option of sum type against rhs"));
				auto unification = unify(option, b, env, bindings, depth + 1);
				if (unification.result) {
					if (unification.bindings.size() > bindings.size()) {
						debug_above(2, log(log_info, "replacing bindings %s with %s",
									str(bindings).c_str(),
									str(unification.bindings).c_str()));
					}
					bindings = unification.bindings;
					return {true, option->str(bindings), bindings};
				} else {
					reasons.push_back(unification.reasons);
				}
			}
			return {false, join(reasons, "\n\t"), {}};
		} else {
			assert(pts_b != nullptr);
			for (auto inbound_option : pts_b->options) {
				debug_above(7, log("checking inbound %s against lhs %s",
							inbound_option->repr().c_str(), a->repr().c_str()));
				auto unification = unify(a, inbound_option, env, bindings, depth + 1);
				if (unification.result) {
					bindings = unification.bindings;
				} else {
					return {
						false,
						string_format(
								"\n\tcould not find a match for \n\t\t%s"
								"\n\tin\n\t\t%s",
								inbound_option->str(bindings).c_str(),
								a->str(bindings).c_str()),
						bindings};
				}
			}
			return {true, "inbound type is a subset of outbound type", bindings};
		}
	} else if (pto_a != nullptr) {
		debug_above(7, log(log_info, "checking inbound type_operator %s",
					pto_a->str().c_str()));
		if (pto_b != nullptr) {
			debug_above(7, log(log_info, "checking outbound type_operator %s",
						pto_b->str().c_str()));
			auto unification = unify(pto_a->oper, pto_b->oper, env, bindings, depth + 1);
			if (unification.result) {
				bindings = unification.bindings;

				if ((pto_a->operand == nullptr) != (pto_b->operand == nullptr)) {
					return {
						false,
						string_format("type mismatch: %s != %s",
								a->str(bindings).c_str(), b->str(bindings).c_str()),
						{}};
				}

				assert(pto_a->operand != nullptr && pto_b->operand != nullptr);

				debug_above(7, log("matching type operands"));
				return unify(pto_a->operand, pto_b->operand, env, bindings, depth + 1);
			}
		} else {
			/* fallthrough, and try expanding the left-hand side */
			debug_above(7, log(log_info, "falling through"));
		}
		auto operator_a = pto_a->oper->rebind(bindings);
		auto operand_a = pto_a->operand->rebind(bindings);

		debug_above(7, log(log_info, "eval_apply(%s, %s, ...)",
					operator_a->str().c_str(), operand_a->str().c_str()));

		/* apply the bindings first, so as to simplify the application */
		auto new_a = eval_apply(operator_a, operand_a, env);

		if (new_a != nullptr) {
			debug_above(7, log(log_info, "eval_apply(%s, %s, ...) -> %s",
						operator_a->str().c_str(), operand_a->str().c_str(),
						new_a->str(bindings).c_str()));

			return unify(new_a, b, env, bindings, depth + 1);
		} else {
			/* types don't match */
			return {
				false,
				string_format("%s <> %s with attempted bindings %s",
						a->str().c_str(),
						b->str().c_str(),
						str(bindings).c_str()),
				{}};
		}
	} else if (ptr_a != nullptr) {
		if (ptr_b != nullptr) {
			debug_above(7, log("matching ptr types"));
			return unify(ptr_a->element_type, ptr_b->element_type, env, bindings, depth + 1);
		} else if (b->is_nil()) {
			if (dyncast<const types::type_managed_t>(ptr_a->element_type) == nullptr) {
				/* managed types cannot take nil, because they are guarded by the maybe type */
				return {true, "", bindings};
			} else {
				return {
					false,
						string_format("%s <> %s with attempted bindings %s because managed types cannot receive nil, unless they are guarded by a maybe (in other words, use a ? after the left-hand-side type name)",
								a->str().c_str(),
								b->str().c_str(),
								str(bindings).c_str()),
						{}};
			}
		} else {
			return {
				false,
					string_format("%s <> %s with attempted bindings %s",
							a->str().c_str(),
							b->str().c_str(),
							str(bindings).c_str()),
					{}};
		}
	} else {
		/* types don't match */
		return {
			false,
			string_format("%s <> %s with attempted bindings %s",
					a->str().c_str(),
					b->str().c_str(),
					str(bindings).c_str()),
			{}};
	}
}

bool unifies(
		types::type_t::ref a,
		types::type_t::ref b,
		const types::type_t::map &env)
{
    unification_t unification = unify(a, b, env);
    return unification.result;
}

File: unification.py
#!/usr/bin/env python
# coding=utf8

"""
Unification for Algebraic Data Types with Macros.
Copyright (c) 2016 William Bradley
The MIT License

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

Attribution: Aspects of this implementation were inspired by Robert
Smallshire's implementation of Andrew Forrest's Scala implementation of
Hindley/Milner Algorithm W.
"""

import json
import os
import copy

UNREACHABLE = 'void'


def error(x):
    return "\033[31m%s\033[0m" % x


def success(x):
    return "\033[32m%s\033[0m" % x


def yellow(x):
    return "\033[33m%s\033[0m" % x


def log(msg, level=1):
    dbg_level = os.environ.get("DEBUG", 0)
    if int(dbg_level) >= int(level):
        print(msg)


class TypeId:

    def __init__(self, name):
        assert isinstance(name, str)
        self.name = name

    def __str__(self):
        assert not "Must have bindings context."

    def to_str(self, bindings):
        return self.name

    def __repr__(self):
        return "TypeId('%s')" % self.name

    def rebind(self, bindings):
        instance = bindings.get(self.name, None)
        if instance:
            return instance.rebind(bindings)
        return self


global __var_counter
__var_counter = 0


def fresh_var():
    global __var_counter
    try:
        return "__%d" % __var_counter
    finally:
        __var_counter += 1


class TypeVariable:

    def __init__(self, name=None):
        if name is None:
            name = fresh_var()
        assert isinstance(name, str)
        self.name = name

    def __str__(self):
        assert not "Must have bindings context."

    def to_str(self, bindings):
        instance = bindings.get(self.name, None)
        if instance:
            return instance.to_str(bindings)
        else:
            return "(any %s)" % self.name

    def __repr__(self):
        return "TypeVariable(%r)" % self.name

    def rebind(self, bindings):
        instance = bindings.get(self.name, None)
        if instance:
            return instance.rebind(bindings)
        return self


class TypeOperator:

    def __init__(self, operator, operand):
        assert not isinstance(operator, str)
        self.operator = operator
        self.operand = operand

    def __str__(self):
        assert not "Must have bindings context."

    def __repr__(self):
        return "TypeOperator(%r, %r)" % (self.operator, self.operand)

    def to_str(self, bindings):
        assert self.operator

        if self.operand:
            return "(%s %s)" % (self.operator.to_str(bindings),
                                self.operand.to_str(bindings))
        else:
            return self.operator.to_str(bindings)

    def rebind(self, bindings):
        return TypeOperator(self.operator.rebind(bindings),
                            self.operand.rebind(bindings))


class TypeProduct:

    def __init__(self, dimensions):
        self.dimensions = dimensions

    def __str__(self):
        assert not "Must have bindings context."

    def __repr__(self):
        return "TypeProduct(%r)" % self.dimensions

    def to_str(self, bindings):
        assert self.dimensions is not None

        return "(and %s)" % ' '.join(
            dimension.to_str(bindings) for dimension in self.dimensions)

    def rebind(self, bindings):
        return TypeProduct([dimension.rebind(bindings)
                            for dimension in self.dimensions])


class TypeSum:

    def __init__(self, options):
        self.options = options

    def __str__(self):
        assert not "Must have bindings context."

    def __repr__(self):
        return "TypeSum(%r)" % self.options

    def to_str(self, bindings):
        assert self.options is not None

        return "(or %s)" % ' '.join(
            option.to_str(bindings) for option in self.options)

    def rebind(self, bindings):
        return TypeSum([option.rebind(bindings)
                        for option in self.options])


def prune(t, bindings):
    """Follow the links across the bindings to reach the final binding."""
    if isinstance(t, TypeVariable):
        if t.name in bindings:
            return prune(bindings[t.name], bindings)
        else:
            return t
    return t


def occurs_in_type(v, type2, bindings):
    """See whether a type variable occurs in a type expression."""
    pruned_type2 = prune(type2, bindings)

    if pruned_type2 == v:
        return True
    elif isinstance(pruned_type2, TypeOperator):
        return occurs_in_type(v, pruned_type2.operand, bindings)

    # TODO: also check inside of TypeSums and TypeProducts

    return False


def unify_terms(env, outbound, inbound):
    """Unifies type terms.
    Returns whether we succeeded, the type, and the bindings used to
    achieve that type.
    """
    # Start by converting the type terms into operators and variables, to
    # simplify things for unification.
    log("unifying terms %s <- %s" % (outbound, inbound), 3)

    outbound_type = outbound.evaluate(env).get_type()
    inbound_type = inbound.evaluate(env).get_type()

    ret, details, bindings = unify(outbound_type, inbound_type, env, {})

    if ret:
        details = outbound_type.to_str(bindings)

    return ret, details, bindings


def no_cycles(unify):
    """Prevent cycles during unification."""
    _visited = {}

    def wrapper(outbound_type, inbound_type, env, bindings):
        log("= %s\n- %s\n" % (outbound_type.to_str(bindings),
                              inbound_type.to_str(bindings)), 2)

        pruned_a = prune(outbound_type, bindings)
        pruned_b = prune(inbound_type, bindings)

        a = pruned_a
        b = pruned_b

        # memoize and do cycle detection with bound terms stringified
        params = "%s|%s|%r" % (
            a.to_str(bindings),
            b.to_str(bindings),
            bindings)

        if params in _visited:
            # memoize and detect recurring calls
            if type(_visited[params]) == bool:
                print("saw already %s" % _visited[params])
                assert False

            return _visited[params]
        else:
            _visited[params] = False
        _visited[params] = unify(outbound_type, inbound_type, env, bindings)
        return _visited[params]
    return wrapper


@no_cycles
def unify(outbound_type, inbound_type, env, bindings):
    # print("Unifying %r with %r" % (outbound_type, inbound_type))
    a = prune(outbound_type, bindings)
    b = prune(inbound_type, bindings)

    if a.to_str(bindings) == b.to_str(bindings):
        return True, '', bindings

    if isinstance(a, TypeVariable):
        if a != b:
            if occurs_in_type(a, b, bindings):
                return False, "recursive unification", bindings

            assert a.name not in bindings

            bindings = bindings.copy()
            bindings[a.name] = b

        return True, '', bindings

    elif isinstance(b, TypeVariable):
        return unify(b, a, env, bindings)
    elif isinstance(a, TypeProduct):
        if not isinstance(b, TypeProduct):
            return (False, "inbound type is not a product type", {})

        if len(a.dimensions) != len(b.dimensions):
            return (False, "product type lengths do not match", {})

        for a_dim, b_dim in zip(a.dimensions, b.dimensions):
            ret, reason, new_bindings = unify(a_dim, b_dim, env, bindings)
            if not ret:
                return ret, reason, {}
            bindings = new_bindings

        return (True, 'products match', bindings)

    elif isinstance(a, TypeSum):
        # the outbound type has options, let's make sure that all possible
        # types in the inbound type can be matched to the outbound type.
        if not isinstance(b, TypeSum):
            reasons = []
            for option in a.options:
                ret, reason, new_bindings = unify(option, b, env, bindings)
                if ret:
                    assert new_bindings is not None
                    bindings = new_bindings

                    # the inbound type matches this option
                    return (
                        ret, 'matches %s' % option.to_str(bindings), bindings)
                reasons.append(reason)

            return (
                False, "inbound type not in the polymorph - \n\t%s" % (
                    '\n\t'.join(reasons)), {})

        else:
            assert isinstance(b, TypeSum)

            # the inbound type also has options! all inbound type possibilities
            # must be unifiable against the outbound options.
            for inbound_option in b.options:
                log("checking inbound %r against %r" % (
                    inbound_option, a), 8)
                ret, reason, new_bindings = unify(a, inbound_option, env,
                                                  bindings)
                if ret:
                    # unification succeeded, accept the bindings
                    bindings = new_bindings
                else:
                    return (
                        ret,
                        "\n\tcould not find a match for \n\t\t%s"
                        "\n\tin\n\t\t%s" % (
                            inbound_option.to_str(bindings),
                            a.to_str(bindings)),
                        bindings)

            assert bindings is not None

            return True, 'inbound type is a subset of outbound type', bindings

    elif isinstance(a, TypeOperator):
        if isinstance(b, TypeOperator):
            # Two type operators, first try to unify the operands
            ret, reason, new_bindings = unify(a.operator, b.operator, env,
                                              bindings)
            if ret:
                assert new_bindings is not None

                if bool(a.operand is None) != bool(b.operand is None):
                    return (
                        False,
                        "Type mismatch: %s != %s" % (a.to_str(new_bindings),
                                                     b.to_str(new_bindings)),
                        {})

                if a.operand and b.operand:
                    return unify(a.operand, b.operand, env, new_bindings)
                else:
                    return True, '', new_bindings

        # attempt expanding the left side
        new_a = eval_apply(a.operator, a.operand, env, bindings)
        if new_a:
            return unify(new_a, b, env, bindings)
        else:
            return (
                False, "%s <> %s /* eval */" % (a.operator.to_str(bindings),
                                     b.operator.to_str(bindings)), {})

    else:
        return (
            False, "%s <> %s" % (a.to_str(bindings),
                                 b.to_str(bindings)), {})


class Lambda:
    __slots__ = ['var', 'body']

    def __init__(self, var, body):
        self.var = var
        self.body = body

    def __repr__(self):
        return "Lambda(%r, %r)" % (self.var, self.body)


def eval_apply(operator, operand, env, bindings):
    assert isinstance(operator, TypeId)

    # See if there is a function in the environment for this operator
    fn = env.get(operator.name)

    if fn:
        # Bind the function's parameter name to the operand
        bindings = copy.copy(bindings)

        assert isinstance(fn.var, TypeId)
        print("Binding %r to %r" % (fn.var.name, operand))
        bindings[fn.var.name] = operand

        return fn.body.rebind(bindings)

    else:
        return None


def main():
    env = {
        'Node': Lambda(TypeId('T'),
                       TypeProduct([
                           TypeId('T'),
                           TypeOperator(TypeId('List'), TypeId('T'))])),
        'List': Lambda(TypeId('T'),
                       TypeSum([
                           TypeId('Null'),
                           TypeOperator(TypeId('Node'), TypeId('T'))])),
        'Maybe': Lambda(TypeId('T'),
                       TypeSum([
                           TypeId('Empty'),
                           TypeOperator(TypeId('Just'), TypeId('T'))])),
    }

    NodeInt = TypeOperator(TypeId('Node'), TypeId('int'))
    MaybeInt = TypeOperator(TypeId('Maybe'), TypeId('int'))
    AnyInt = TypeOperator(TypeVariable(), TypeId('int'))
    ProdIn = TypeProduct([TypeId('int'), NodeInt])

    # ListInt = TypeOperator(TypeId('List'), TypeVariable())
    tv = TypeVariable()
    ListListInt = TypeProduct([
        tv,
        TypeOperator(TypeVariable(), TypeId('int'))])

    result, reason, bindings = unify(MaybeInt, TypeId('Empty'), env, {})
    print(result)
    print(reason)
    print(bindings)
    print("Final type = %s" % MaybeInt.rebind(bindings).to_str(bindings))


def bindings_to_str(bindings):
    return json.dumps(bindings, sort_keys=True, indent=4)


if __name__ == '__main__':
    main()
